src/aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -405,21 +406,34 @@ def handle_known_unrecoverable_failure(self, calculation):
         exit_codes=[
             PwCalculation.exit_codes.ERROR_COMPUTING_CHOLESKY,
             PwCalculation.exit_codes.ERROR_DIAGONALIZATION_TOO_MANY_BANDS_NOT_CONVERGED,
+            PwCalculation.exit_codes.ERROR_S_MATRIX_NOT_POSITIVE_DEFINITE,
+            PwCalculation.exit_codes.ERROR_ZHEGVD_FAILED,
+            PwCalculation.exit_codes.ERROR_QR_FAILED,
+            PwCalculation.exit_codes.ERROR_EIGENVECTOR_CONVERGENCE,
+            PwCalculation.exit_codes.ERROR_BROYDEN_FACTORIZATION,
         ]
     )
     def handle_diagonalization_errors(self, calculation):
         """Handle known issues related to the diagonalization.
 
-        Switch to ``diagonalization = 'cg'`` if not already running with this setting, and restart from the charge
-        density. In case the run already used conjugate gradient diagonalization, abort.
+        We use the following strategy. When a diagonalization algorithm fails, we try using an other one
+        still not used. Conjugate gradient (CG) is kept as last option, as it is the slowest among the
+        available ones, but on the contrary it is the most stable as well, thus kept as `last resort`.
+
+        If also CG is failing, abort.
         """
-        if self.ctx.inputs.parameters['ELECTRONS'].get('diagonalization', None) == 'cg':
+        current = self.ctx.inputs.parameters['ELECTRONS'].get('diagonalization', 'david')
```

Comment:
Is the reason for choosing `david` as the default in line 429 because it is the default in QE? Has that always been the case for all versions? Do we parse this from the output? If so, it might be safer to get the current value from the parsed outputs if not explicitly defined in the input parameters

Reply:
I am rather confident `david` is the default for all the versions, or for sure the versions the plugin is supporting right now. But I'll check the output, and take that one in case

Reply:
If it is not in the outputs, fine to just hard-code it then for now